File: content/examples/other/leetcode/problems/binary-tree-level-order-traversal/example-2/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
  static class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
      this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
      this.val = val;
      this.left = left;
      this.right = right;
    }
  }

  static List<List<Integer>> levelOrderDFS(TreeNode root) {
    List<List<Integer>> out = new ArrayList<>();
    dfs(root, 0, out);
    return out;
  }

  static void dfs(TreeNode node, int depth, List<List<Integer>> out) {
    if (node == null) return;
    if (out.size() == depth) out.add(new ArrayList<>());
    out.get(depth).add(node.val);
    dfs(node.left, depth + 1, out);
    dfs(node.right, depth + 1, out);
  }

  public static void main(String[] args) {
    TreeNode root = new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));
    System.out.println(levelOrderDFS(root));
  }
}

File: content/examples/other/leetcode/problems/binary-tree-level-order-traversal/example-2/solution.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function levelOrderDFS(root) {
  const out = [];
  (function dfs(node, depth) {
    if (!node) return;
    if (out.length === depth) out.push([]);
    out[depth].push(node.val);
    dfs(node.left, depth + 1);
    dfs(node.right, depth + 1);
  })(root, 0);
  return out;
}

if (require.main === module) {
  const root = new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));
  console.log(levelOrderDFS(root));
}

module.exports = { TreeNode, levelOrderDFS };
